Sources/Yams/Parser.swift

```diff
@@ -355,13 +355,32 @@ private extension Parser {
             pairs.append((key, value))
             event = try parse()
         }
+        let keys = pairs.map { $0.0 }
+        try checkDuplicates(mappingKeys: keys)
         let node = Node.mapping(.init(pairs, tag(firstEvent.mappingTag), event.mappingStyle, firstEvent.startMark))
         if let anchor = firstEvent.mappingAnchor {
             anchors[anchor] = node
         }
         return node
     }
 
+    private func checkDuplicates(mappingKeys: [Node]) throws {
+        var duplicates: [String: [Node]] = [:]
+        for key in mappingKeys {
```

Comment:
this for loop could just be
```
let duplicateKeys = Dictionary(grouping: keys, by: { $0.string! }).filter { $1.count > 1 }.keys
```
but I wanted to avoid the force unwrap. if we're confident that these keys will always be scalars (which I think they _should_ be but I'm not sure about that) then I can just switch to that simpler form

Reply:
What about? 
```suggestion
        let duplicateKeys = Dictionary(grouping: keys) { 
                                                                $0.string ?? "[non-scalar-key]
                                                            }
                                                            .filter { $1.count > 1 }
```

But actually Node.Mapping has a .string value too via Node.string -> String.construct(from:)

And actually [Yaml.org](https://yaml.org/spec/1.2.2/#742-flow-mappings) 
Header: "Example 7.21 Single Pair Implicit Entries" seems to indicate that non-scalar keys are allowed?

So perhaps what would be most correct here here was to put the nodes in a Set and see if the set has fewer elements than the keys list. 

Or Yams could just document the behavioral deficiency. It's pretty weird to use a key that's not a scalar. 😆

Reply:
oh, interesting! didn't realize that having non-scalar keys was allowed. I like the idea of using a `Set`, but after giving implementing this a try I realized that the existing structure makes it a lot easier to print a helpful error message when this issue pops up. 

That said, given that `Node`s are `Hashable`, I can just use them as the dictionary keys. This works since the metadata isn't used in `hash(into:)`, only the tag and value ([eg for Node.Scalar](https://github.com/jpsim/Yams/blob/main/Sources/Yams/Node.Scalar.swift#L91)) (otherwise the same key in a different line would hash to a different value). Updated and the tests I wrote earlier (back in July, wow!) continue to succeed so I think this method works better. Thanks for the suggestion!